File: 113_path_sum_ii.java
class Solution {
    private List<List<Integer>> solution; 
    private List<Integer> candidate;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        solution = new ArrayList<>();
        candidate = new ArrayList<Integer>();

        this.tryExpand(root, targetSum);

        return this.solution;
    }

    public void path(TreeNode root, int remSum) {
        if(root.left == null && root.right == null && remSum == 0) 
            this.solution.add(new ArrayList(this.candidate)); 

        this.tryExpand(root.left, remSum);
        this.tryExpand(root.right, remSum);
    }

    public void tryExpand(TreeNode root, int remSum) {
        if(root != null) {
            this.candidate.add(root.val);
            remSum -= root.val;
            this.path(root, remSum);
            this.candidate.remove(candidate.size()-1);
        }
    }
}

File: 113_path_sum_ii.py
class Solution(object):
    def pathSum(self, root, targetSum):
        return [s[::-1] for s in self.path(root, targetSum)]
    
    def path(self, root, remSum):
        if not root:
            return []
        if root.left == None and root.right == None and remSum-root.val == 0:
            return [[root.val]]

        leftCand = path(root.left, remSum - root.val)
        rightCand = path(root.right, remSum - root.val)
        
        return [path + [root.val] for path in leftCand + rightCand]